sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -65,13 +68,25 @@ private static DoFnSignature parseSignature(Class<? extends DoFn> fnClass) {
     TypeToken<?> inputT = null;
     TypeToken<?> outputT = null;
 
-    // Extract the input and output type.
+    PCollection.IsBounded isBounded = null;
+
+    // Extract the input and output type, and whether the fn is bounded.
     checkArgument(
         DoFn.class.isAssignableFrom(fnClass),
         "%s must be subtype of DoFn",
         fnClass.getSimpleName());
     TypeToken<? extends DoFn> fnToken = TypeToken.of(fnClass);
     for (TypeToken<?> supertype : fnToken.getTypes()) {
+      if (supertype.getRawType().isAnnotationPresent(DoFn.Bounded.class)) {
+        checkArgument(
+            isBounded == null, "Both @Bounded and @Unbounded specified on %s", formatType(fnToken));
+        isBounded = PCollection.IsBounded.BOUNDED;
+      }
+      if (supertype.getRawType().isAnnotationPresent(DoFn.Unbounded.class)) {
+        checkArgument(
+            isBounded == null, "Both @Bounded and @Unbounded specified on %s", formatType(fnToken));
```

Comment:
I think at some point we'll want this to use machinery like `AvroDeterminismChecker` (https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java#L362) which encapsulates the validation code and allows recording multiple problems before failing rather than only reporting the first problem.

That should probably be done in a separate PR.